drivers/adc/adc_stm32.c

```diff
@@ -626,69 +629,86 @@ static void adc_stm32_unset_common_path(const struct device *dev, uint32_t PathI
 	LL_ADC_SetCommonPathInternalCh(__LL_ADC_COMMON_INSTANCE(adc), PathInternal);
 }
 
-static void adc_stm32_teardown_channels(const struct device *dev, uint8_t channel_id)
+static void adc_stm32_teardown_channels(const struct device *dev)
 {
 	const struct adc_stm32_cfg *config = dev->config;
+	struct adc_stm32_data *data = dev->data;
 	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
 
+	for (uint8_t channel_id = 0; channel_id < STM32_CHANNEL_COUNT; channel_id++) {
+		if (data->channels & BIT(channel_id)) {
 #ifdef CONFIG_SOC_SERIES_STM32G4X
-	if (config->has_temp_channel) {
+			if (config->has_temp_channel) {
 #if DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay)
-		if ((__LL_ADC_CHANNEL_TO_DECIMAL_NB(LL_ADC_CHANNEL_TEMPSENSOR_ADC1) == channel_id)
-		    && (config->base == ADC1)) {
-			adc_stm32_disable(adc);
-			adc_stm32_unset_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
-		}
+				if ((__LL_ADC_CHANNEL_TO_DECIMAL_NB(
+					     LL_ADC_CHANNEL_TEMPSENSOR_ADC1) == channel_id) &&
+				    (config->base == ADC1)) {
+					adc_stm32_disable(adc);
```

Comment:
This (and all the other copy-pasta of this same code block) will call `adc_stm32_disable(adc);` over and over in the loop.  Why not move the disable call outside the loop?

IMHO, a function that does a (channel_id, adc_id) →LL_ADC_PATH_constant mapping would make this code and the copy of it in setup_channels massively smaller and easier to read.  The ifdefs are out of control.

Reply:
Ageed. If you do, please use a specific commit.

Reply:
I moved the `adc_stm32_disable` to before the loop, and decrease the indentation by one. IMO this is is much more readable, and could maybe be cleaned up if required later when this driver is made more readable.

Reply:
I've decided that this is out of scope for this PR, and can be cleanup up in the inevitable refactor/cleanup of this driver. For now I've moved the `disable` back into the `if`s so that it's not called unnecessarily.